src/System.Device.Gpio/System/Device/I2c/I2cController.cs

```diff
@@ -0,0 +1,218 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+using System.Collections.Generic;
+
+namespace System.Device.I2c
+{
+    public sealed partial class I2cController : II2cController
+    {
+        private readonly Dictionary<(int, int), I2cDevice> _openDevices;
+
+        /// <summary>
+        /// Initializes new instance of I2cController.
+        /// </summary>
+        public I2cController()
+        {
+            _openDevices = new Dictionary<(int, int), I2cDevice>();
+        }
+
+        /// <summary>
+        /// Gets the I2C device for specified address.
+        /// </summary>
+        /// <param name="busId">The bus ID the I2C device is connected to.</param>
+        /// <param name="address">The bus address of the I2C device to get.</param>
+        /// <returns>The I2C device with specified address.</returns>
+        private I2cDevice GetDevice(int busId, int address)
+        {
+            if (!_openDevices.TryGetValue((busId, address), out I2cDevice device))
+            {
+                throw new InvalidOperationException("The specified bus ID and address is not being used with an I2C device.");
+            }
+
+            return device;
+        }
+
+        /// <summary>
+        /// Opens an I2C device with specified settings in order for it to be ready to use with the controller.
+        /// </summary>
+        /// <param name="settings">The connection settings of a device on an I2C bus.</param>
+        /// <param name="shouldDispose">Determines if the device should be disposed by a controller or binding.</param>
+        public void OpenDevice(I2cConnectionSettings settings, bool shouldDispose = true)
```

Comment:
there is something confusing about this name/API. Usually `Open` methods return whatever they try to open and never dispose it.

Reply:
I was trying to use similar naming between controllers.  See [OpenPin for GpioController](https://github.com/dotnet/iot/blob/33432162e8b7dbc64f457571a9be43767033df51/src/System.Device.Gpio/System/Device/Gpio/GpioController.cs#L63) where it doesn't return anything.

The ShouldDispose is to allow controllers or bindings to know if they should dispose of device when they are being removed.

Something for future thought (and understanding it might not be feasible)...  We could eventually modify the GpioController to manage GpioPin (or GpioDevice) and update  to something like...

```csharp
public void OpenPin(int pin, bool shouldDispose = true)
// Or...
public void OpenPin(GpioConnectionSettings settings, bool shouldDispose = true)
```

Where this could create default GpioPin and offer a flag to dispose of pin when done with controller or binding.  This might address your issue #347.  But I'm not completely sure.

Reply:
the main difference is that pin currently does not have a representing object vs I2cDevice does.

I'm personally not in favor of keeping track of all open I2c/Spi devices just to add few helper functions for edge case scenarios

Reply:
Agreed, but would be an object if GPIO was expanded upon.  Same thing for possibly changing PwmController.  Currently, we manage chips/channels.  We could expand and manage PwmDevices as you mentioned in #204.

```csharp
public void OpenChannel(int chip, int channel, bool shouldDispose = true)
// Or...
public void OpenChannel(PwmConnectionSettings settings, bool shouldDispose = true)
```

Reply:
I think I'm leaning toward changing every `Open*` method to return something which can be disposed and remove all `Close*` functions. I think with the current path we will get ourselves into really annoying state. I'd leave all helper functions as they are (WritePin etc).

In general I'd prefer if we did not track lifetime of objects in their parents at all but shouldDispose is acceptable to me. cc: @joperezr 

Initially I was really pushing back on having any objects for pins since it seemed heavyweight to me but I think I don't feel that way anymore consdering all special cases we've hit (i.e. #347 and weird abstraction for PWM).

If we're gonna go this path we should change it before we ship 3.0. I think we should make the changes mostly source compatible (so we can consider leaving the Close methods although I'd prefer if we've removed them)

Reply:
Not included any longer.